MekHQ/src/mekhq/service/MassRepairService.java

```diff
@@ -780,61 +1013,131 @@ private static boolean checkArmorSupply(IPartWork part) {
+					debugLog(
+							"...... ending calculateNewMassRepairWorktime with previousWorkTime due time reduction skill mod now being less that required skill - %s ns",
+							System.nanoTime() - time);
+
+					return new WorkTimeCalculation(previousNewWorkTime);
+				}
+			}
+
+			// If we have a null newWorkTime, we're done. Use the previous one.
 			if (null == newWorkTime) {
+				debugLog("...... ending calculateNewMassRepairWorktime because newWorkTime is null - %s ns",
+						System.nanoTime() - time);
+
 				if (!increaseTime) {
 					return new WorkTimeCalculation(previousNewWorkTime);
-				} else {
-					WorkTimeCalculation wtc = new WorkTimeCalculation();
-
-					Skill skill = tech.getSkillForWorkingOn(part);
+				}
 
-					if (skill.getExperienceLevel() >= highestAvailableTechSkill) {
-						wtc.setReachedMaxSkill(true);
-					}
+				WorkTimeCalculation wtc = new WorkTimeCalculation();
 
-					return wtc;
+				if (skill.getExperienceLevel() >= highestAvailableTechSkill) {
+					wtc.setReachedMaxSkill(true);
 				}
+
+				return wtc;
 			}
 
-			if (part instanceof Part) {
-				((Part) part).setMode(newWorkTime);
+			// Set our new workTime and calculate the new targetRoll
+			if (partWork instanceof Part) {
+				((Part) partWork).setMode(newWorkTime);
 			}
 
-			targetRoll = campaign.getTargetFor(part, tech);
+			targetRoll = campaign.getTargetFor(partWork, tech);
 
+			// If our roll is impossible, revert to the previous one
 			if ((targetRoll.getValue() == TargetRoll.IMPOSSIBLE) || (targetRoll.getValue() == TargetRoll.AUTOMATIC_FAIL)
 					|| (targetRoll.getValue() == TargetRoll.CHECK_FALSE)) {
-				continue;
+				debugLog("...... ending calculateNewMassRepairWorktime due to impossible role - %s ns",
+						System.nanoTime() - time);
+
+				return new WorkTimeCalculation(previousNewWorkTime);
 			}
 
 			if (increaseTime) {
+				// If we've reached our BTH, kick out. Otherwise we'll loop
+				// around again
 				if (targetRoll.getValue() <= mro.getBthMin()) {
+					debugLog(
+							"...... ending calculateNewMassRepairWorktime because we have reached our BTH goal - %s ns",
+							System.nanoTime() - time);
+
 					return new WorkTimeCalculation(newWorkTime);
 				}
 			} else {
 				if (targetRoll.getValue() > mro.getBthMax()) {
+					debugLog(
+							"...... ending calculateNewMassRepairWorktime because we have reached our BTH goal - %s ns",
+							System.nanoTime() - time);
+
 					return new WorkTimeCalculation(previousNewWorkTime);
-				}
+				} else if (targetRoll.getValue() > mro.getBthMax()) {
+					debugLog(
+							"...... ending calculateNewMassRepairWorktime because we have reached our BTH goal - %s ns",
+							System.nanoTime() - time);
 
-				return new WorkTimeCalculation(newWorkTime);
+					return new WorkTimeCalculation(newWorkTime);
+				}
 			}
 		}
 
 		return new WorkTimeCalculation();
 	}
 
+	private static void debugLog(String msg, Object... replacements) {
+		if (!debugMode) {
+			return;
+		}
```

Comment:
Instead of using a separate debugMode variable, you can use `if (!MehKQ.getLogger().willLog(getClass(), LogLevel.DEBUG) {` for the same purpose.  This way, if the _log4j.xml_ config gets updated to change the logging level on this class, you're already accounted for.

Reply:
Sure. That's fine. As long as someone doesn't turn on DEBUG as the root. The MRMS debug logging is pretty spammy.